accepted/0000-dependency-selector-syntax.md

```diff
@@ -0,0 +1,421 @@
+- [`:is(<selector list>)`](https://developer.mozilla.org/en-US/docs/Web/CSS/:is)
+- [`:root`](https://developer.mozilla.org/en-US/docs/Web/CSS/:root) matches the root node/dependency
+- [`:scope`](https://developer.mozilla.org/en-US/docs/Web/CSS/:scope) matches node/dependency it was queried against
+- [`:empty`](https://developer.mozilla.org/en-US/docs/Web/CSS/:empty) when a dependency has no dependencies
+- [`:private`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#private) when a dependency is private
+- `:link` when a dependency is linked
+- `:deduped` when a dependency has been deduped
+- `:override` when a dependency is an override
+- `:extraneous` when a dependency could be but is not deduped
+- `:outdated` when a dependency is not `latest`
+- `:vulnerable` when a dependency has a `CVE` 
+- `:semver(<spec>)` matching a valid [`node-semver`](https://github.com/npm/node-semver) spec
+- `:path(<path>)` [glob](https://www.npmjs.com/package/glob) matching based on dependencies path relative to the project
+- `:realpath(<path>)` [glob](https://www.npmjs.com/package/glob) matching based on dependencies realpath
+- `:type(<type>)` [based on currently recognized types](https://github.com/npm/npm-package-arg#result-object)
+
+#### Generic Pseudo Selectors:
+
+A standardized pseudo selector pattern will be used for `Object`s, `Array`s or `Arrays` of `Object`s accessible via `Arborist`'s `Node.package` metadata. Pseudo classes generated in this way will allow for iterative attribute selection.
+
+`Array`s specifically use a special `value` keyword in place of a typical attribute name. `Arrays` also support exact `valu`e matching when a `String` is passed to the selector. See examples below:
+
+#### Example of an `Object`: 
+```css
+/* return dependencies that have a `scripts.test` containing `"tap"` */
+*:scripts([test~=tap])
+```
+
+#### Example of an `Array` Attribute Selection:
+```css
+/* return dependencies that have a keyword that begins with "react" */
+*:keywords([value^="react"])
+```
+
+#### Example of an `Array` matching directly to a value:
+```css
+/* return dependencies that have the exact keyword "react" */
+/* this is equivalent to `*:keywords([value="react"])` */
+*:keywords("react")
+```
+
+#### Example of an `Array` of `Object`s: 
+```css
+/* returns */
+*:contributors([email="ruyadorno@github.com"])`
+```
+
+### Classes
+
+Given that `Arborist` will control our understanding of the DOM (Dependency Object Model), claseses are predefined by their relationships to their ancestors & are specific to their dependency type. Dependencies may have, & are allowed to have, multiple classifications (ex. a `workspace` may also be a `dev` dependency).
+
+- `.prod`
+- `.dev`
+- `.optional`
+- `.peer`
+- `.bundled`
+- `.workspace`
+
+### Attributes
+
+There are several known attributes that are normalized & queryable living in `Node.package` (aka. `package.json`).
```

Comment:
is there a reason this isn't generic? For example, TypeScript has the [`volta`](https://github.com/microsoft/TypeScript/blob/main/package.json#L120) property in their package.json. It might be nice to be able to dig into other packages that have `volta` as well.

Reply:
Generics are explained [here[(https://github.com/npm/rfcs/blob/e63094a66d3b690561c129599861f221ffb7305c/accepted/0000-dependency-selector-syntax.md#generic-pseudo-selectors). It may not be apparent (which we should fix) but the intent is that any information in a dependencies `package.json` will be queryable - there's just a subset of those fields that are "known" to us from the outset (aka. https://docs.npmjs.com/cli/v8/configuring-npm/package-json).

Querying for `volta` specific keys/values would look something like:

```bash
npm query "*[volta]"
npm query "*:volta([node=14.15.5])"
```

Reply:
npm already conceptually knows about `type`, and `exports`, and `browser`, and webpack's `sideEffects`, and a bunch of others, but these aren't in the list.

Reply:
Reworking attributes here in an update